### DISCUSSION TITLE
Feature Request: Private properties

### DISCUSSION THREAD

**Discussion by saschaisele-zf** [eclipse-edc/Connector#2592] [Ideas]
### Feature Request
Extend the following entities with private properties which are not exposed outside the Management API:
Assets, Policy Definitions, Contract Definitions, Contract Negotiations, Contract Agreements, Transfer Process

### Which areas would be affected and how?
**SPI Data Types**
Add a field named 'privateProperties' to the following entities:
- Asset
- Policy
- Contract Definition
- Contract Negotiations
- Contract Agreements
- Transfer Process

**APIs, Models and Transformers**
Allow reading and manipulation of the 'privateProperties' in the following APIs:

Management API
- Asset API
- Policy Definition API
- Contract Definition API
- Contract Negotiations API
- Contract Agreements API
- Transfer Process API

This includes adjusting existing models and transformers to accommodate the addition of the 'privateProperties'.

### Why is this feature desired?
Enriching these entities with information not relevant for external consumers/providers is currently not possible.
To achieve this, it has to be done through an external storage.

### Example:
Additional tags related to choosing policies for an asset are added to the private properties.
They govern which policies are recommended for the asset and help the user with the creation of contract offers.
Since those tags represent meta-information meant for internal use, there is no benefit to be had in sending this information with the asset inside the contract offer.
On the contrary, there is the possibility that an external consumer/provider may gleam information about internal processes which represents a potential risk.

### Point of discussion:
How to contain the availability of 'privateProperties' to the Management API.

**Comment by jimmarino** [comment]
@saschaisele-zf can you ping me so we can discuss at some point? 

I understand the use cases and motivations for private properties on Assets and ContractDefinitions, but I am unsure about the other entities. We also need to keep in mind a few things:

1. This work would need to be staged after/in coordination with the introduction of JSON-LD in milestone 8 
2. We should create a Decision Record
3. Each entity that needs private properties should be handled in a separate Issue and PR (it's OK to have a master issue)
4. Private properties need to work with JSON-LD namespaces (I can explain what is involved here)

**Comment by mohannad-ezzo** [comment]
Hello @jimmarino ,
Regarding the contract definition entity, we come to a usecase (Storing administrative data like created by, created at, changed by,  changed at) where the private properties are required for the contract definition entity. Therefore, we extended the contract definition entity with the private properties map and enabled the CRUD operations functionalities via management APIs of contract definition.

The corresponding changes are ready, tested and we would like to contribute them. The changes, I did, are being done in the same way of implementation for the private properties of Asset entity (which is available as of now)
Can I raise PR for the same?

**Reply by jimmarino** [reply]
Before you submit a PR, can you please raise an [adoption request](https://github.com/eclipse-edc/Connector/issues/new?assignees=&labels=adoption&projects=&template=adoption_request.md&title=Adoption+Request) discussing the specific motivations?

Overall, this seems consistent with other entities such as Assets, so it sounds reasonable.

**Comment by suh-rao** [comment]
Hello @jimmarino ,
Regarding the policy definition entity, we come to a usecase (Storing administrative data like created by, created at, changed by, changed at, policyType as access or usage) where the private properties are required for the policy definition entity. Therefore, we extended the policy definition entity with the private properties map and enabled the CRUD operations functionalities via management APIs of policy definition.

The corresponding changes are ready, tested and we would like to contribute them. The changes, I did, are being done in the same way of implementation for the private properties of Asset entity (which is available as of now)
Can I raise PR for the same?
Adaption Request: https://github.com/eclipse-edc/Connector/issues/3567

**Reply by ndr-brt** [reply]
Please raise the PR, but we won't merge it if the feature was implemented in the same way as how it's implemented now for the Asset entity, please adapt to how has been done for `TransferProcess`/`ContractDefinition`, also `Asset` will adapt to the same design (#3559 )

**Reply by suh-rao** [reply]
sure will do it as per  TransferProcess/ContractDefinition
